extensions/s3/src/main/java/io/deephaven/extensions/s3/S3ChannelContext.java

```diff
@@ -121,22 +114,27 @@ int fill(final long position, final ByteBuffer dest) throws IOException {
             final int impliedReadAhead = (int) (lastFragmentIx - firstFragmentIx);
             final int desiredReadAhead = instructions.readAheadCount();
             final long totalRemainingFragments = numFragments - firstFragmentIx - 1;
-            final int maxReadAhead = instructions.maxCacheSize() - 1;
-            readAhead = Math.min(
-                    Math.max(impliedReadAhead, desiredReadAhead),
-                    (int) Math.min(maxReadAhead, totalRemainingFragments));
+            readAhead = Math.min(Math.max(impliedReadAhead, desiredReadAhead), totalRemainingFragments);
         }
-        final S3Request firstRequest = getOrCreateRequest(firstFragmentIx);
+        // Hold a reference to the first request to ensure it is not evicted from the cache
+        S3Request.AcquiredRequest acquiredFirstRequest = getOrCreateRequest(firstFragmentIx);
         for (int i = 0; i < readAhead; ++i) {
+            // Do not hold references to the read ahead requests
             getOrCreateRequest(firstFragmentIx + i + 1);
         }
         // blocking
-        int filled = firstRequest.fill(position, dest);
+        int filled = acquiredFirstRequest.request.fill(position, dest);
+        acquiredFirstRequest.release();
+        acquiredFirstRequest = null;
```

Comment:
My motivation was to make this request available for GC once we are done with the fill.
But there was something unexpected happening. Basically, originally I only had 
`acquiredFirstRequest = null;` 
But then, I started seeing a lot of errors indicating the underlying buffer got freed, even though we are holding the handle to the first request in line 120. I think that was the compiler removing the reference to the variable, since it is not being used after this point.

So I added a release, just to indicate to the compiler that we need this variable, so don't clear it.
Please let me know if there is a better way to do this.

Reply:
I think that if you make acquiredFirstRequest an instance member of the context, instead of a local, you'll address the issue. I'm pretty confident that you're right and the compiler is allowing the reference to be collected.